RecoEgamma/EgammaElectronAlgos/interface/EgammaPixelMatchParamObjects.h

```diff
@@ -36,56 +45,161 @@
 	      float phiMin,float phiMax)const{
       return x>=absEtaMin && x<absEtaMax && y>=nrClusMin && y<=nrClusMax 
 	&& z>=phiMin && z < phiMax;
     }
   };
-
+  template<>
+  inline AbsEtaNrClusPhi<reco::ElectronSeed>::AbsEtaNrClusPhi(const reco::ElectronSeed& seed){
+    reco::SuperClusterRef scRef = seed.caloCluster().castTo<reco::SuperClusterRef>();
+    x = std::abs(scRef->eta());
+    y = scRef->clustersSize();
+    z = scRef->phi();
+  }
+  
+  template<typename T>
   struct AbsEtaNrClusEt {
     float x;
     size_t y;
     float z;
 
-    AbsEtaNrClusEt(const reco::ElectronSeed& seed){
-      reco::SuperClusterRef scRef = seed.caloCluster().castTo<reco::SuperClusterRef>();
-      x = std::abs(scRef->eta());
-      y = scRef->clustersSize();
-      z = scRef->energy()*sin(scRef->position().Theta());
+    AbsEtaNrClusEt(const T& seed){
+      x = std::abs(seed.eta());
+      y = seed.nrClus();
+      z = seed.et();
     }
     bool pass(float absEtaMin,float absEtaMax,size_t nrClusMin,size_t nrClusMax,
 	      float etMin,float etMax)const{
       return x>=absEtaMin && x<absEtaMax && y>=nrClusMin && y<=nrClusMax 
-	&& z>=etMin && z < etMax;
+	&& z>=etMin && (z < etMax || etMax<0);
+    }
+  };
+  template<>
+  inline AbsEtaNrClusEt<reco::ElectronSeed>::AbsEtaNrClusEt(const reco::ElectronSeed& seed){  
+    reco::SuperClusterRef scRef = seed.caloCluster().castTo<reco::SuperClusterRef>();
+    x = std::abs(scRef->eta());
+    y = scRef->clustersSize();
+    z = scRef->energy()*sin(scRef->position().Theta());
+  };
+  
+  template<typename T>
+  struct EtAbsEta {
+    float x;
+    float y;
+    
+    EtAbsEta(const T& seed){
+      x = seed.et();
+      y = std::abs(seed.eta());
+    }
+    bool pass(float etMin,float etMax,float absEtaMin,float absEtaMax){
+      return  x>=etMin && (x < etMax || etMax<0) && y>=absEtaMin && y<absEtaMax;
     }
   };
   
+  template<>
+  inline EtAbsEta<reco::ElectronSeed>::EtAbsEta(const reco::ElectronSeed& seed){  
+    reco::SuperClusterRef scRef = seed.caloCluster().castTo<reco::SuperClusterRef>();
+    x = scRef->energy()*sin(scRef->position().Theta());
+    y = std::abs(scRef->eta());
+  };
+    
+  template<typename T>
+  struct EtAbsEtaCharge {
+    float x;
+    float y;
+    int z;
+    
+    EtAbsEtaCharge(const T& seed){
+      x = seed.et();
+      y = std::abs(seed.eta());
+      z = seed.charge();
+    }
+    bool pass(float etMin,float etMax,float absEtaMin,float absEtaMax,int chargeMin,int chargeMax){
+      return  x>=etMin && (x < etMax || etMax<0) && y>=absEtaMin && y<absEtaMax && z>=chargeMin && z<=chargeMax;
+    }
+  };  
+  template<>
+  inline EtAbsEtaCharge<reco::ElectronSeed>::EtAbsEtaCharge(const reco::ElectronSeed& seed){  
+    reco::SuperClusterRef scRef = seed.caloCluster().castTo<reco::SuperClusterRef>();
+    x = scRef->energy()*sin(scRef->position().Theta());
+    y = std::abs(scRef->eta());
+    z = seed.getCharge();
+  } 
+  template<typename T>
+  struct EtAbsEtaPhi {
```

Comment:
is this variety of multiplexed selections going to continue to proliferate?

It seems more practical to define .AND. and .OR. operations and simplify the current implementation and the possible future extensions.